番号付きページ送り機能

sd-tomita opened the pull request:
## 概要

今までのページャーはこうだった
```
[←前へ]  [次へ→]
```

こういうのを作れるようにしたい。
```
[←前へ] [1][2][3][4][5] [次へ→]
```
## TODO

※随時追加
- [x] Pager にページ情報を持つ内部クラスを追加する
- [x] PagerLink に上記内部クラスを使ってリンクタグを作るメソッドを追加する
- [x] マニュアル追加
- [x] 総ページ数より多い数を返せるのを修正する
- [x] 動作チェックなど
- レビュー後修正作業
  - [x] PageHolderの名前を変更
  - [x] メンバ変数をプロパティに変更
  - [x] 不要な三項演算子をカット
  - [x] `<span>` タグに付与するクラス属性の変更
  - [x] GetLinksTag にコールバックを渡せるようにする
  - [x] PageData クラスのプロパティは get 以外はアクセス制限する
  - [x] GetLinksTag メソッドの重複コードを排除する
  - [x] current_page ---> current に変更

sd-tomita committed:
wip

sd-tomita committed:
ユニットテストを追加

まだGetPagesを実装していないのでビルドはできません

sd-tomita committed:
ページ番号と現在のページかどうかを管理する内部クラスを追加

sd-tomita committed:
if 文の判定が無駄に呼ばれるのを修正

sd-tomita committed:
Sdx.Html.PagerLink のテスト用クラスを追加

sd-tomita committed:
GetLinksTag のテストケースを追加

sd-tomita committed:
GetLinksTag メソッドを実装

sd-tomita committed:
番号付きページャーの生成例をドキュメントに追加

sd-tomita committed:
取得可能なページ数より多い数を指定した場合のテストを追加

結果、案の定コケた

sd-tomita committed:
最大ページ数より多い数を指定されても取得可能な分だけを返すように修正

sd-tomita commented:
## 動作チェック

概ね問題なさそう
- 基本は現在のページを真ん中に
- 左か右に必要なページが足りないときは真ん中に出さないで、寄せるだけ
- 番号リンクの数が総ページ数を超えることはない。

**全7ページ、ページリンクは5件まで出す場合の表示遷移**

| 現在のページ | 表示状態 |
| --- | --- |
| 1 | `1 [2][3][4][5]` |
| 2 | `[1] 2 [3][4][5]` |
| 3 | `[1][2] 3 [4][5]` |
| 4 | `[2][3] 4 [5][6]` |
| 5 | `[3][4] 5 [6][7]` |
| 6 | `[3][4][5] 6 [7]` |
| 7 | `[3][4][5][6] 7` |

**ページリンクを10件に指定する**
ページは全部で7ページなのでそれ以上はリンクは出ない。
遷移は以下のとおり。

| 現在のページ | 表示状態 |
| --- | --- |
| 1 | `1 [2][3][4][5][6][7]` |
| 2 | `[1] 2 [3][4][5][6][7]` |
| 3 | `[1][2] 3 [4][5][6][7]` |
| 4 | `[1][2][3] 4 [5][6][7]` |
| 5 | `[1][2][3][4] 5 [6][7]` |
| 6 | `[1][2][3][4][5] 6 [7]` |
| 7 | `[1][2][3][4][5][6] 7` |

gomo commented on Sdx/Pager.cs:
```diff
@@ -144,5 +144,46 @@ public bool HasPrev
       }
     }
+
+    public class PageHolder
```
既にPageがあったらPageHolderにしたんだと思うけど、この単語からは複数のページを持ってるイメージがわきますね。何にしようか・・・・困りましたね。PageDataかな。

😇

gomo commented on Sdx/Pager.cs:
```diff
@@ -144,5 +144,46 @@ public bool HasPrev
+    public class PageHolder
+    {
+      public bool is_current;
+      public int number;
```
これはpublicクラスなのでプロパティにしたほうがいいですね。

それぞれ、IsCurrent/Id　がいいかな。Pager内ので扱いと合わせてPageData.Pageというのも考えたけえどちょいきもいね。

Pager.PageId/Page.id

にすべきでしたね。結構使ってるので今更ですけど。

😇

gomo commented on Sdx/Pager.cs:
```diff
@@ -144,5 +144,46 @@ public bool HasPrev
+
+    public bool HasPage(int page)
+    {
+      return (page >= 1 && page <= LastPage) ? true : false;
```
これ、三項演算子にする意味ある？`return page >= 1 && page <= LastPage`で良さそうな気がした。直すほどでもないけど、ついでなので整理してもらおうかな。

😇

gomo commented on UnitTest/Sdx/Html/PagerLink.cs:
```diff
@@ -0,0 +1,34 @@
+      var pagerLink = new Sdx.Html.PagerLink(pager, baseUrl, "pid");
+
+      var links = pagerLink.GetLinksTag(5);
+      Assert.Equal("<span class=\"disabled\">11</span>", links[2].Render());
```
これは、`disabled`だとまずそうですね。current的なものがいいかな。

😇

gomo commented on Sdx/Pager.md:
```diff
@@ -42,3 +42,28 @@ select.LimitPager(pager);
+    <div class="link-list">
+      <%-- GetLinksTag の引数分だけリンクタグが生成されます  --%>
+      <% foreach(var link in pager_link.GetLinksTag(5)) { %>
+        <%= link.Render() %>
```
この流れだとaタグの中身がページの数字固定になっちゃって使い勝手が悪そうです。
`pager_link.GetLinksTag(5, page => string.format("[{0}]", page))`みたいにコールバックを渡せるようにしてください。コールバックがなかったら数字のみでOKです。

😇

sd-tomita committed:
クラス名を変更

PageHolder は中に複数持っているイメージを与えうるため。

sd-tomita committed:
メンバ変数をプロパティに変更

public なので、このクラスのメンバ変数に
直接アクセスできてしまうのは良くないため。
現状はpublicのようなものだが、プロパティにしておくことで
何かしたいときあとでプロパティ内部の変更だけで済む。

sd-tomita committed:
コメント修正

sd-tomita committed:
不要な三項演算子を削除

結局 bool が返るのでわざわざ三項演算子にする必要なかった。

sd-tomita committed:
クラス名修正時の漏れを対応

sd-tomita committed:
ユニットテスト修正

sd-tomita committed:
現在のページ部分にdisableのクラス属性が付いていたのを変更

ひとまずタグを直接作って対応

sd-tomita committed:
コールバックを渡すテストを追加

sd-tomita committed:
CreateLinkTag 後にクラス名を変えれば済むことだったので変更

わざわざ自前で作らなくても、既存のメソッド使ったあとに
ちょこっと手を加えれば済むのでそれで対応

sd-tomita committed:
GetLinksTag にコールバックを渡せるオーバーロードを追加

sd-tomita committed:
current 時に付与するクラス名を切り替えできるようにしました

オーバーロード追加することも考えましたがメンバ変数足すほうが早かったので。

sd-tomita committed:
コールバックにHTMLのタグを入れたときエンコードされていたので修正

これだとWEBページ上に'<b>11</b>'という文字列が出てしまうので
今回はHTMLエンティティ化をfalseにしました。

gomo commented on Sdx/Pager.cs:
```diff
@@ -147,8 +147,17 @@ public bool HasPrev
+      {
+        get;
+        set;
+      }
```
あ、先に行っておけばよかったけど、このsetはinternalにできそうじゃない？難しそう？
できたらお願いします。

😇

sd-tomita replied:
なるほど、たしかに。
set だけなら問題なくいけそうです。ユニットテスト通れば internal でコミットします！

gomo commented on Sdx/Html/PagerLink.cs:
```diff
@@ -118,9 +120,24 @@ public Tag GetLast()
+        var linkTag = CrateLinkTag(pd.IsCurrent ? null : pd.Id.ToString());
+        linkTag.AddText(func(pd.Id.ToString()), false);
         links.Add(linkTag);
       });
```
この二つの関数はオーバーロードではなく、デフォルト値を使ってIFで分岐するか、委譲するかすると重複を排除可能ですね。

## 委譲

``` c#
public List<Tag> GetLinksTag(int number)
{
  return GetLinksTag(number, num => num);
}

public List<Tag> GetLinksTag(int number, Func<string, string> func)
{
  classNameForDisabled = "current_page";
  var links = new List<Tag>() { };
  Pager.GetPageDataList(number).ForEach(pd => {
    var linkTag = CrateLinkTag(pd.IsCurrent ? null : pd.Id.ToString());
    linkTag.AddText(func(pd.Id.ToString()), false);
    links.Add(linkTag);
  });

  return links;
}
```

## デフォルト値

``` c#
public List<Tag> GetLinksTag(int number, Func<string, string> func = null)
{
  classNameForDisabled = "current_page";
  var links = new List<Tag>() { };
  Pager.GetPageDataList(number).ForEach(pd => {
    var linkTag = CrateLinkTag(pd.IsCurrent ? null : pd.Id.ToString());
    linkTag.AddText(func != null ? func(pd.Id.ToString()) : pd.Id.ToString(), false);
    links.Add(linkTag);
  });

  return links;
}
```

サニタイズは数字の時はしなくてもOKなのでどっちにしてもfalseでOKかと

どっちがいいかな。委譲のほうがかっこいい？wかっこいいは冗談として、気持ち委譲の方が負荷が高そうですけど、わかりやすいのは委譲の方ですね。負荷は微々たるもんなので委譲でいきましょう。

基本、オーバーロードした時は委譲ができないか検討してください。親玉のメソッドがあって追加処理をして親玉に委譲するイメージですかね。`One fact in one place`の原則ですね。

またオーバーロードは引数のデフォルト値に書き換えが可能で、デフォルト値ほうがわかりやすい場合もあるかも。

😇

sd-tomita replied:
> 気持ち委譲の方が負荷が高そうですけど、わかりやすいのは委譲の方ですね。負荷は微々たるもんなので委譲でいきましょう。

委譲がコードの見通し良くてわかりやすいですね。委譲で修正しますー

> またオーバーロードは引数のデフォルト値に書き換えが可能で、デフォルト値ほうがわかりやすい場合もあるかも。

言われてみればPHPで書くときはデフォルト値よく使ってますね。
了解です、次回以降はその都度検討してわかりやすい方を使っていこうと思います。

gomo commented on Sdx/Html/PagerLink.cs:
```diff
@@ -118,9 +120,24 @@ public Tag GetLast()
+
+    public List<Tag> GetLinksTag(int number, Func<string, string> func)
+    {
+      classNameForDisabled = "current_page";
```
ふむ・・・直すほどじゃないけど、`disabled`と`current_page`っておかしくないですか？`disabled_link`と`current_page`か`disable`と`current`かって感じがしました・・・細かすぎ？ 😄 

既に使用されているので`current`がいいかな。

sdx内のことだし、他を直すついでがあるので直してもらおうかな。

😇

sd-tomita committed:
現在のページ分のタグに付与するクラス名を変更

現在のページ以外が disabled という一単語なので
対応関係を合わせました。

sd-tomita committed:
重複コードの排除

sd-tomita committed:
PageData クラスプロパティの set メソッドアクセス修飾子変更

private だと Pager クラスからも操作できなくなるので
internal で対応してます。

sd-tomita committed:
コールバックを使う場合の記述例をドキュメントに追加

gomo commented:
👍 
おっけー！こいつはmasterにマージでOKかな。
